### ISSUE TITLE
unable to move window under wayland

### ISSUE THREAD

**Issue by mendaxcat** [kovidgoyal/kitty#4802]
**Describe the bug**
after commit `ac407d42`, the window cannot be moved after launching kitty

**To Reproduce**
1. launch kitty using `kitty --config NONE`
2. use mouse to drag the window
3. the window simply does not move

**Environment details**
```
kitty 0.24.4 (90ed5959de) created by Kovid Goyal
Linux mendaxca 5.10.0-11-amd64 #1 SMP Debian 5.10.92-1 (2022-01-18) x86_64
Debian GNU/Linux 11 mendaxca /dev/tty

Running under: Wayland
Frozen: False
Paths:
  kitty: /tmp/kitty/linux-package/bin/kitty
  base dir: /tmp/kitty/linux-package/lib/kitty
  extensions dir: /tmp/kitty/linux-package/lib/kitty/kitty
  system shell: /bin/bash

Config options different from defaults:

Important environment variables seen by the kitty process:
	PATH                                /home/mendaxca/bin:/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games
	LANG                                C.UTF-8
	VISUAL                              nvim
	SHELL                               /bin/bash
	GLFW_IM_MODULE                      ibus
	DISPLAY                             :0
	WAYLAND_DISPLAY                     wayland-0
	USER                                mendaxca
	XDG_MENU_PREFIX                     gnome-
	XDG_SESSION_DESKTOP                 gnome
	XDG_SESSION_TYPE                    wayland
	XDG_CURRENT_DESKTOP                 GNOME
	XDG_SESSION_CLASS                   user
	XDG_RUNTIME_DIR                     /run/user/1000
```

**Comment by Nefsen402** [comment]
That's odd that surface commits (usually meant to resize) is effecting the moving the window. I guess this must be a problem with client side decorations, correct? According to the bisect we're talking about a commit that just moves the `wl_surface_commit` to comply more with the wayland protocol which is really unexpected! Maybe we alway want to run this after the surface commit.
```
inform_compositor_of_window_geometry(window, "configure");
if (live_resize_done) _glfwInputLiveResize(window, false);
```

**Comment by kovidgoyal** [comment]
Yeah it's likely to be a CSD issue. I spend as little time on Wayland as I can possibly get away with, so not sure what the appropriate fix is. Does the wayland protocol actually require the surface to be committed even if the configure didnt change anything? Seems rather wasteful.

**Comment by Nefsen402** [comment]
`inform_compositor_of_window_geometry` will set the xdg_surafce geometry bounds. This state is double buffered meaning it gets sent in the next wl_surface_commit. In fact this means that the old code was wrong, because it would commit, then send the geometry. However it seems to be breaking things nontheless. `live_surface_done` should be false when just moving the window so there should be no functional difference for ac407d42dee2f52d3d8eaa24ecb5cd0e357e27cb. Are we absolutely sure it's this commit? I think we're dealing with a compositor bug here though.

**Comment by Nefsen402** [comment]
Does this patch fix the problem for you?

```
diff --git a/glfw/wl_window.c b/glfw/wl_window.c
index b98d0a6f..c65e64dd 100644
--- a/glfw/wl_window.c
+++ b/glfw/wl_window.c
@@ -519,10 +519,10 @@ static void xdgSurfaceHandleConfigure(void* data,
         debug("final window content size: %dx%d\n", window->wl.current.width, window->wl.current.height);
         _glfwInputWindowFocus(window, window->wl.current.toplevel_states & TOPLEVEL_STATE_ACTIVATED);
         ensure_csd_resources(window);
-        inform_compositor_of_window_geometry(window, "configure");
         if (live_resize_done) _glfwInputLiveResize(window, false);
     }
 
+    inform_compositor_of_window_geometry(window, "configure");
     wl_surface_commit(window->wl.surface);
 }
 ```

**Comment by mendaxcat** [comment]
moving the `inform_compositor_of_window_geometry()` does not fix the problem

__some additional info__
* the window is positioned on top-left of the screen without title bar after launch
* the window will work as expected after maximizing
![image](https://user-images.githubusercontent.com/90375986/157273959-4712e52f-e4a3-4be2-bf8e-5e4bc868346c.png)

**Comment by Nefsen402** [comment]
Hmm, it looks like CSD is not working at all. Only after entering/exiting fullscreen you start seeing decorations that you can interact with and move? This is what it looks like through tinywl which seems to handle the current kitty release correctly.
![image](https://user-images.githubusercontent.com/3621017/157276043-7220daf4-bb01-4653-a0db-d814153d6f77.png)

**Comment by Nefsen402** [comment]
I stared at the code and I noticed that decorations weren't being handled correctly much like toplevel configures. (https://wayland.app/protocols/xdg-decoration-unstable-v1#zxdg_toplevel_decoration_v1:event:configure). I made a patch that should be more conformant. Does this patch fix things?
[patch.txt](https://github.com/kovidgoyal/kitty/files/8207855/patch.txt)

**Comment by mendaxcat** [comment]
yes, window decorations appear after entering and exiting full screen

i tried building `tinywl`, but the decorations are still missing in it. then i installed another wayland compositor called `weston`, kitty(v0.24.4 with or without patch) works fine in it. however, the decorations are still not shown under gnome with or without you last patch.

**Comment by kovidgoyal** [comment]
> @kovidgoyal Any reason you added the set flag for toplevel configures? It shouldn't be necessary. I understand that the toplevel configure might come in the wrong order if you're running on a broken compositor, but if that were the case then the condition checking for width/height/toplevel_states would fail.

I like to be explicit about these things. The intended behavior is that commit should occur after a configure, the set flag ensures that is always the case, regardless of other conditions, which might change in the future.

**Comment by Nefsen402** [comment]
I found numerous problems while investigating this further. CSD has a lot of nuance and some of the problems I fixed definitely will explain this issue. 
[patch.txt](https://github.com/kovidgoyal/kitty/files/8212911/patch.txt)

**Comment by kovidgoyal** [comment]
I will note that this issue doesn't reproduce for me with current master under GNOME 41. The only issues I can find are the long standing ones of right clicking on the titlebar being rather iffy, sometime th emenu disappears immediately sometimes it appears in the wrong place.

**Comment by Nefsen402** [comment]
Yeah, I haven't been able to reproduce this exact issue as well. I hate to be passing random patches to people but it would be nice to get this fixed.

**Comment by kovidgoyal** [comment]
Hmm well #4797 is an issue I can reproduce but that your patch doesn't fix. I am fairly sure it is also a regression caused by the change to the configure event handling, as I remember testing this very scenario when developing the code.

**Comment by Nefsen402** [comment]
I bisected issue #4797 to the same commit that this issue was complaining about and I figured out all that was going on. Apparently, `dispatchChangesAfterConfigure` will resize the egl surface, but will not actually spin up the the render loop. Apparently, GNOME does not like it if we `wl_surface_commit` a surface that has been resized, but nothing has been rendered to it yet. I fixed it by adding a check that will not call `wl_surface_commit` in handle_configure if a frame has been scheduled. In this case, we'll just wait for the frame callback which will end up calling `wl_surface_commit` anyway. If the surface has not been resized, then it will still call `wl_surface_commit` as per the requirement in the wayland spec. That one was really annoying to figure out. Amended patches:
[patch.txt](https://github.com/kovidgoyal/kitty/files/8214405/patch.txt)

I'll wait for @mendaxcat to tell us the verdict of his issue, then I'll put together the PR. Thanks @kovidgoyal for being so patient!

**Comment by kovidgoyal** [comment]
On Wed, Mar 09, 2022 at 04:00:54AM -0800, Alexander Orzechowski wrote:
> [patch.txt](https://github.com/kovidgoyal/kitty/files/8214405/patch.txt)
> 
> I'll wait for @mendaxcat to tell us the verdict of his issue, then I'll put together the PR. Thanks @kovidgoyal for being so patient!

No worries. Thanks for your efforts! I dont use Wayland so I am grateful
to have someone else deal with issues in it.

**Comment by mendaxcat** [comment]
i am sorry to inform you that this issue persists after applying your latest patch in my environment

debian stable (gnome 3.38.5) still has this issue. fedora 35 and debian testing (gnome 42) do not have this problem. maybe there are some problems with debian 11

**Comment by Nefsen402** [comment]
What happens when you force server side decorations off? On line 515 in glfw/wl_window.c put the line `window->wl.decorations.serverSide = false`. That should always render decorations.

**Comment by mendaxcat** [comment]
unfortunately this does not help

**Comment by kovidgoyal** [comment]
Your patch fixes the rezise on blur issue for me on GNOME 41.

**Comment by Nefsen402** [comment]
I'll put up a PR later today or tomorrow. I'm busy now. But I'm unsure how to fix mendaxcat's issue.

**Comment by kovidgoyal** [comment]
On Wed, Mar 09, 2022 at 04:00:54AM -0800, Alexander Orzechowski wrote:
> I bisected issue #4797 to the same commit that this issue was complaining about and I figured out all that was going on. Apparently, `dispatchChangesAfterConfigure` will resize the egl surface, but will not actually spin up the the render loop. Apparently, GNOME does not like it if we `wl_surface_commit` a surface that has been resized, but nothing has been rendered to it yet. I fixed it by adding a check that will not call `wl_surface_commit` in handle_configure if a frame has been scheduled. In this case, we'll just wait for the frame callback which will end up calling `wl_surface_commit` anyway. If the surface has not been resized, then it will still call `wl_surface_commit` as per the requirement in the wayland spec. That one was really annoying to figure out. Amended patches:
> [patch.txt](https://github.com/kovidgoyal/kitty/files/8214405/patch.txt)
> 
> I'll wait for @mendaxcat to tell us the verdict of his issue, then I'll put together the PR. Thanks @kovidgoyal for being so patient!

I merged your PR but am sad to say that I can now reproduce @mendaxcat's
issue on GNOME 41. But only when running kitty on a scale 2 monitor and
launching it from the activities bar not from another terminal. So it
looks timing related. I think there is probably something the GNOME
compositor doesn't like about how we damage and commit the CSD surfaces.

I have found that when the problem is reproduced, the compositor is
partially unaware that the CSD surfaces exist. They are rendered, but
dont yield pointer enter/leave events, which is why they cannot be used
to drag the window. Resizing the window (by pressing super+left and
super+down) fixes the issue from then on.

**Comment by Nefsen402** [comment]
Does this help the situation at all:
```
diff --git a/glfw/wl_window.c b/glfw/wl_window.c
index 782d3692..ecfc57e0 100644
--- a/glfw/wl_window.c
+++ b/glfw/wl_window.c
@@ -516,11 +516,6 @@ static void xdgSurfaceHandleConfigure(void* data,
         int width = window->wl.pending.width, height = window->wl.pending.height;
         set_csd_window_geometry(window, &width, &height);
         resized = dispatchChangesAfterConfigure(window, width, height);
-        if (window->wl.decorations.serverSide) {
-            free_csd_surfaces(window);
-        } else {
-            ensure_csd_resources(window);
-        }
         debug("final window content size: %dx%d resized: %d\n", width, height, resized);
     }
 
@@ -528,6 +523,12 @@ static void xdgSurfaceHandleConfigure(void* data,
 
     if (!resized) wl_surface_commit(window->wl.surface);
     window->wl.pending_state = 0;
+
+    if (window->wl.decorations.serverSide) {
+        free_csd_surfaces(window);
+    } else {
+        ensure_csd_resources(window);
+    }
 }
 
 static const struct xdg_surface_listener xdgSurfaceListener = {
```
Perhaps GNOME has problems with attaching subsurfaces on an unmapped surfaces. It should be fine either way.
If this doesn't help, try removing the `if (!resized) wl_surface_commit(window->wl.surface);`. Maybe gnome has issues for subsurfaces on non rendered brand new egl surface.

**Comment by kovidgoyal** [comment]
Nope, neither of those worked. Shot in the dark: We currently damage and
commit the CSD in ensure_csd_resources(). Maybe move that commit to
happen only when commiting the main surface.

**Comment by Nefsen402** [comment]
Foot will instantiate its CSD subsurfaces same as we are https://codeberg.org/dnkl/foot/src/commit/6e536e7ed8f1de2615fa323123ddb2b06a45a179/wayland.c#L745. The only difference is that they are acking the configure after all that. Maybe that's what all this is about we just need to move the ack down to line 526 or something.

**Comment by kovidgoyal** [comment]
Nope putting the ack before     

inform_compositor_of_window_geometry(window, "configure");

or after it or at the end of the function made no difference. And the
issue doesnt repro for me with foot, just as a point of interest.

**Comment by Nefsen402** [comment]
I'm still having trouble reproducing this on my setup. If it's really caused by timing, it would help if I could get some `--debug-rendering` logs when the CSD is glitched.

**Comment by kovidgoyal** [comment]
The logs are identical with and without the glitch so its timing in
mutter not kitty, but here is the log with the glitch launched form
activities:

```
GL version string: '4.6 (Core Profile) Mesa 21.3.7' Detected version: 4.6
CSD: old.size: 0x0 new.size: 640x400 needs_update: 1 size_changed: 1
Created decoration buffers at scale: 1 vertical_height: 436 horizontal_width: 664
top-level configure event: size: 0x0 states:
Resizing framebuffer to: 640x400 at scale: 2
CSD: old.size: 640x400 new.size: 640x400 needs_update: 1 size_changed: 1
Created decoration buffers at scale: 2 vertical_height: 436 horizontal_width: 664
final window content size: 640x400 resized: 1
Setting window geometry in configure event: x=0 y=-24 640x424
top-level configure event: size: 640x424 states: TOPLEVEL_STATE_ACTIVATED
CSD: old.size: 640x400 new.size: 640x400 needs_update: 1 size_changed: 0
final window content size: 640x400 resized: 0
Setting window geometry in configure event: x=0 y=-24 640x424
Resizing framebuffer to: 640x400 at scale: 2
CSD: old.size: 640x400 new.size: 640x400 needs_update: 0 size_changed: 0
CSD: old.size: 640x400 new.size: 640x400 needs_update: 0 size_changed: 0
prompt_marking: x=0 y=0 op='A'
CSD: old.size: 640x400 new.size: 640x400 needs_update: 0 size_changed: 0
```

And here is the log without the glitch launched form gnome-terminal
```
GL version string: '4.6 (Core Profile) Mesa 21.3.7' Detected version: 4.6
CSD: old.size: 0x0 new.size: 640x400 needs_update: 1 size_changed: 1
Created decoration buffers at scale: 1 vertical_height: 436 horizontal_width: 664
top-level configure event: size: 0x0 states:
Resizing framebuffer to: 640x400 at scale: 2
CSD: old.size: 640x400 new.size: 640x400 needs_update: 1 size_changed: 1
Created decoration buffers at scale: 2 vertical_height: 436 horizontal_width: 664
final window content size: 640x400 resized: 1
Setting window geometry in configure event: x=0 y=-24 640x424
top-level configure event: size: 640x424 states: TOPLEVEL_STATE_ACTIVATED
CSD: old.size: 640x400 new.size: 640x400 needs_update: 1 size_changed: 0
final window content size: 640x400 resized: 0
Setting window geometry in configure event: x=0 y=-24 640x424
Resizing framebuffer to: 640x400 at scale: 2
CSD: old.size: 640x400 new.size: 640x400 needs_update: 0 size_changed: 0
CSD: old.size: 640x400 new.size: 640x400 needs_update: 0 size_changed: 0
prompt_marking: x=0 y=0 op='A'
CSD: old.size: 640x400 new.size: 640x400 needs_update: 0 size_changed: 0
```

**Comment by kovidgoyal** [comment]
Would you like the WAYLAND_DEBUG logs as well. They are useless to  me as I am not familiar enough with wayland an dmutter, but maybe they will help you,

**Comment by Nefsen402** [comment]
The WAYLAND_DEBUG logs certainly wouldn't hurt. However, I don't think they would tell me anything I don't know. I have one question: Do the drop shadows work while in the bugged state and only the title bar is not visible, or no CSD shows up at all?

**Comment by mendaxcat** [comment]
According to <https://github.com/kovidgoyal/kitty/issues/4802#issuecomment-1064137215>, I removed Line 531 (`if (!resized) wl_surface_commit(window->wl.surface);`) of `glfw/wl_window.c` based on <https://github.com/kovidgoyal/kitty/commit/14e0b01b4085c54e05ae7ba778fe63939be5b9e2>. Then the problem disappears.

**Comment by Nefsen402** [comment]
`if (!resized) wl_surface_commit(window->wl.surface);` will just change the timing of these things and may not make the problem reproduce as kovidgoyal did not have success with removing that call. I think this was a long standing issue, it's just that some recent changes made the problem more obvious.

**Comment by kovidgoyal** [comment]
On Fri, Mar 11, 2022 at 04:37:08AM -0800, Alexander Orzechowski wrote:
> The WAYLAND_DEBUG logs certainly wouldn't hurt. However, I don't think they would tell me anything I don't know. I have one question: Do the drop shadows work while in the bugged state and only the title bar is not visible, or no CSD shows up at all?

OK will generate it in a little while. It varies, on first start all of
them are usually partially rendered. After that switching to another
window and back fixes the rendering. But the compositor still doesnt
send pointer events for the regions. Its almost like the buffers are not
being sent fully to the compositor. Or the damage regions are not
correct.

**Comment by kovidgoyal** [comment]
I found I can get mutter to behave by simply ignoring the first 0x0 configure event it sends. I have gated the workaround to apply only when XAUTHORITY contains mutter so hopefully it wont break anything else.

**Comment by real-or-random** [comment]
Fwiw, the specific workaround introduced here bites me with the following config:
 - GNOME / Mutter 42.4
 - Kitty 0.25.2
 - `hide_window_decorations yes`
 - scaling factor 2 

When I run kitty, GNOME "assumes" the window is only half the size in vertical and horizontal (can be seen when looking at the highlighted window frame when doing Alt+Tab), and then sometimes the window gets resized to the smaller config.  

This doesn't happen when I launch kitty with `XAUTHORITY= kitty`, so I assume it's relevant here.

Sorry to report this, I know this has a long history....

My current workaround is to set `remember_window_size no` and run kitty with `--start-as=fullscreen` (which I want anyway). That specific combination seems to trigger the resizing in a way such that the problem disappears but seems rather a coincidence, and it works only for users who want a fullscreen window.

**Comment by real-or-random** [comment]
> My current workaround is to set `remember_window_size no`

Some more testing that this was a red herring. I see the issue with both `yes` and `no` sometimes, it seems a timing thing.

That is still an issue with Kitty 0.26.4 and GNOME 42.5 and `hide_window_decorations yes` or `XAUTHORITY=` make it go away as a workaround.  Any ideas? Should I open a new issue for this?

**Comment by kovidgoyal** [comment]
I dont see how XAUTHORITY= could make a difference. The XAUTHORITY check was removed

**Comment by real-or-random** [comment]
Ok sorry, this was wrong testing on my side. I was testing in a different terminal with `XAUTHORITY= kitty` but this bypassed my .desktop file, which has `--start-as-fullscreen`. 

After some more testing, I see the following with Kitty 0.26.4 and GNOME 42.5 when `remember_window_size yes` and kitty is invoked with `--start-as=fullscreen`. What seems to happen is that I can see kitty starting with the (small) remembered window size for the fraction of a second and then it switches to fullscreen mode. Then, *only in some invocations of kitty*, when I Alt+Tab after Kitty went to full-screen, mutter assumes the old window size (which can be seen from the thick frame mutter draws in Alt+Tab to indicate the window to switch to.)

Maybe this is some race when going fullscreen already during startup.

**Comment by kovidgoyal** [comment]
Running kitty as

kitty --config NONE --start-as fullscreen

does not repro for me in mutter 42.5
No matter how many times I alt+tab kitty always comes back as full
screen. Also for me alt+tab shows application logo not a border around
the window.

**Comment by real-or-random** [comment]
Sorry for imprecise instructions. When I said "Alt+Tab", I actually meant a customized shortcut for "Switch windows directly". This is how you can get the border. (And the border is just an easy way to spot the window position and dimensions. The quirky behavior goes beyond the border, so this is not a purely cosmetic issue).
 
See the following screencast:

[Screencast from 2022-10-24 10:50:06.webm](https://user-images.githubusercontent.com/1071625/197488167-6fa77b59-aac3-48cc-9e8f-2739bbfd15cf.webm)

 - In the initial frame, you see that extensions are disabled (just to make sure).
 - Then I show how that the "Switch windows directly" shortcut is set to "Super+Tab" (okay, I lied with "Alt"...)
 - Then I try to reproduce the issue by using gnome-terminal to invoke `echo "hide_window_decorations yes" | kitty --config -`, close kitty (which will remember a non-fullscreen window size), and  then invoke `echo "hide_window_decorations yes" | kitty --config - --start-as fullscreen`.) Note that I screw the up the order of events a few times, and also that the bug doesn't show up initially.
 - When you skip to 0:54, you can see a successful reproduction of the issue. The first few window switches are with Super hold for a few second, and I just press tab a few times. The kitty windows is displayed as full screen but the blue border is centered with the original window size.
 - Then at around 1:05, I release Super and press it again to restart the GNOME's window switching thing. At this point, you see that mutter now actually resizes the kitty window to what it thinks is its size (but moves it to the top-left corner). It's a little bit hard to see because both the kitty window (white font) and the gnome-terminal window (green font) have a black background. This shows that the issue is not just that the wrong border is displayed, i.e., it is not just cosmetic as I said above.

**Comment by kovidgoyal** [comment]
I dont use GNOME so how do I configure this special shortcut? And does
the issue happen only when using this method of switching windows? Using
the default method works?

**Comment by real-or-random** [comment]
> I dont use GNOME so how do I configure this special shortcut?

I just noticed that the GNOME defaults for this are Alt+Escape, so you could use that shortcut. Alternatively, run  `gnome-control-center keyboard`, click "View and Customize Shortcuts", search for "Switch windows directly" and set a shortcut.

> And does the issue happen only when using this method of switching windows? Using the default method works?

No, I just reproduced this with the default method (Alt+Tab or Super+Tab). You don't see the border but the fullscreen window will be resized when doing Alt+Tab. 

I also could reproduce it with the
 - `kitty --config NONE`
 -  close kitty (which will remember a non-fullscreen window size)
 - `echo "hide_window_decorations yes" | kitty --config - --start-as fullscreen`

If I remove the `hide_window_decorations yes` entirely, I can't see the bug anymore.

**Comment by kovidgoyal** [comment]
On Tue, Oct 25, 2022 at 01:48:28AM -0700, Tim Ruffing wrote:
> > I dont use GNOME so how do I configure this special shortcut?
> 
> I just noticed that the GNOME defaults for this are Alt+Escape, so you could use that shortcut. Alternatively, run  `gnome-control-center keyboard`, click "View and Customize Shortcuts", search for "Switch windows directly" and set a shortcut.

Yes, with Alt+Esc I can reproduce it, but as far as I can tell it's a
mutter bug. Run kitty like this

kitty --config=NONE --start-as fullscreen -o hide_window_decorations=y
--debug-rendering

>From the debug rendering output you will see that the final size kitty
sends to mutter is the full screen size. Why mutter doesnt respect that, 
I have no idea. Likely some timing issue in mutter? When window
decorations are enabled the sequence of events is timed differently as
the decoration surfaces have to be committed as well. 

I suggest reporting it to mutter maybe they can tell us what the issue
is.

**Comment by real-or-random** [comment]
> I suggest reporting it to mutter maybe they can tell us what the issue is.

Their response is that it's definitively a bug in kitty, see their detailed analysis here:  
https://gitlab.gnome.org/GNOME/mutter/-/issues/2484

**Comment by kovidgoyal** [comment]
I'm afraid that doesnt make any sense. Why should mutter treat changing
to fullscreen any differently from any other resize? If what they are
saying is correct then this problem should happen with any resize not
just a full screen, since for any resize, kitty will first set window
geometry and only sometime later commit a new rendered surface.
Seems completely inconsistent behavior in mutter. 

OK I'll investigate it in more detail when I can find the time. Maybe I
can just delay the surface commit after full screening in GLFW to after
kitty has rendered the resized surface on the next render frame event.
IIRC I already added code for this in GLFW for normal resizes, to satisfy
GNOME, maybe it isnt getting triggered for fullscreens. Although this
solution is really a hack since many other events besides a render frame
can trigger surface commits in glfw.

**Comment by real-or-random** [comment]
>  If what they are
> saying is correct then this problem should happen with any resize not
> just a full screen, since for any resize, kitty will first set window
> geometry and only sometime later commit a new rendered surface.

As I understand them, the correct order is 1. set window geometry, and then 2. commit surface with the same geometry. 
I don't know what kitty does for normal resizes. But at least here, this rule seems to be violated because 1. kitty sets fullscreen geometry but then 2. commits a surface with a non-fullscreen buffer.

> OK I'll investigate it in more detail when I can find the time. 

Ok, thanks, sounds good. Do you want me to open a new issue to better keep track of this? Having it in this closed issue is somewhat inconsistent.

**Comment by jadahl** [comment]
> Why should mutter treat changing to fullscreen any differently from any other resize?

That's not what happens, the issue is that the fullscreen window geometry is committed before the fullscreen surface content matches the window geometry. This will at worst result in a client disconnect if s/fullscreen/maximized/ and strict compositors, or best, glitchy/buggy behavior.

> Maybe I can just delay the surface commit after full screening in GLFW

What Wayland and xdg-shell mandates is that when you receive a xdg_toplevel.configure event with a fullscreen state, you're expected to provide a complete and correct surface state that you acknowledge using xdg_surface.ack_configure and wl_surface.commit. The wl_surface.commit is expected to contain and complete a complete surface state including correctly sied and rendered buffers, and a matching/corresponding window geometry. The issue here is that Kitty first wl_surface.commit's first half, then the second half, which is invalid.

**Comment by kovidgoyal** [comment]
On Thu, Oct 27, 2022 at 05:03:29AM -0700, Jonas Ådahl wrote:
> > Why should mutter treat changing to fullscreen any differently from any other resize?
> 
> That's not what happens, the issue is that the fullscreen window geometry is committed before the fullscreen surface content matches the window geometry. This will at worst result in a client disconnect if s/fullscreen/maximized/ and strict compositors, or best, glitchy/buggy behavior.

Yes, but the exact same thing happens when doing a normal resize. GLFW's
wayland backed does not treat fullscreen and resize any differently. I
did add a workaround to GLFW to not commit the surface until the next
render frame for resizes, precisely for GNOME, so maybe this workaround
is not working for fullscreen, though offhand I dont know why that
would be.

> 
> > Maybe I can just delay the surface commit after full screening in GLFW
> 
> What Wayland and xdg-shell mandates is that when you receive a xdg_toplevel.configure event with a fullscreen state, you're expected to provide a complete and correct surface state that you acknowledge using xdg_surface.ack_configure and wl_surface.commit. The wl_surface.commit is expected to contain and complete a complete surface state including correctly sied and rendered buffers, and a matching/corresponding window geometry. The issue here is that Kitty first wl_surface.commit's first half, then the second half, which is invalid.

That's a pretty painful design constraint, buffers in OpenGL
applications are attached to surfaces on swap. The render loop 
is controlled by the application not the windowing framework (GLFW)
There's no way for GLFW to know when the application will next decide
to swap buffers. This works fine on every platform other than Wayland.

Since kitty has its own private GLFW fork I suppose I can break that
layering and have GLFW hold off on committing the window surface after a
resize until the next swap is done by kitty. Am I correct in thinking
that as long as the window surface is committed only after the resized
buffer is attached to it, mutter will be happy?

And thanks for your input, I dont really know too much about Wayland as
I dont use it, so its quite helpful for me to hear from somebody that
works on a compositor.

**Comment by kovidgoyal** [comment]
On Thu, Oct 27, 2022 at 05:27:57AM -0700, Jonas Ådahl wrote:

> Perhaps one can call it a painful constraint, but it is a necessity to achieve the "every frame is perfect" design goal of Wayland.
> 
> If you first commit a half and incomplete state, then follow up with the rest, if that would be made legal, we would have baked in an imperfect frame if we ended up drawing between those committed state.

Apologies, I dont want to debate the design of Wayland, I obviously dont
know enough about it. Here is the design for kitty's wayland backend I
am proposing, please let me know if it is correct.

1) When xdg top level configures are received, kitty does nothing
2) When the window surface configure is received, kitty first acks it
3) Then it informs the compositor of the new window geometry
4) Then it attaches a blank buffer of the correct size to the window
surface
5) then it commits the surface

This means that there will actually be a flash of blanked pixels
so very much not every frame being perfect, but trying to make every
frame perfect with this design isn't really feasible for an application
that renders independently of window manager interactions.

My question is that apart from the flash of blank pixels that will later
change to the correct window contents on the next render frame, there
are no problems with this design from the perspective of the compositor?

> 
> > Am I correct in thinking that as long as the window surface is committed only after the resized buffer is attached to it, mutter will be happy?
> 
> Yes, what mutter, and all compositors *expect* (ignoring client bug work-arounds, or or disconnecting clients) is that you commit an updated window geometry together with the matching buffer when you ack a configuration, otherwise they cannot possibly present a correct state.

This confused me some more, sorry. Do I have to ack the configuration
only after committing the surface? Currently GLFW acks it before.

**Comment by jadahl** [comment]
> please let me know if it is correct

I can tell you how I'd do it, assuming I have control of when `eglSwapBuffers()` or equivalent is called, and when the first OpenGL draw call is done.

1) When xdg toplevel configure is received - cache the received size and state
2) When xdg surface configure is received - resize the EGLSurface according to the size, and set up your painting pipeline to paint according to the toplevel state, ack the configuration and queue a repaint. Do *not* wl_surface_commit()
3) Paint the frame using OpenGL, let it end with eglSwapBuffers()
4) Make sure there are no wl_surface_commit() before 3)

What this means on the Wayland wire, is that you prepare the state passed from Kitty to the compositor by calling `xdg_surface.set_window_geometry()`, `xdg_surface.ack_configure()`, and then rely on EGL to `wl_surface.attach()` and `wl_surface_commit()` for you. In the end, what the compositor sees is a window geometry, acked configuration, a new correctly painted and sized buffer, all together in a single commit.

This is, more or less, how e.g. GTK4 work.

> This confused me some more, sorry. Do I have to ack the configuration only after committing the surface? Currently GLFW acks it before.


Think of all `wl_surface` and `xdg_surface` state you set to end up in a box of pending state, that the compositor doesn't yet look at. Only when you commit does the compositor open the box to apply it. If you for example ack immediately, as long as you do not commit before you also paint, the ack is in the pending box, and won't be looked at by the compositor until the commit happens, which would be done via EGL.

**Comment by kovidgoyal** [comment]
OK got it, thanks.

**Comment by jadahl** [comment]
There are a couple of possibly badly documented 'gotchas" here too: with EGL, you resize the EGLSurface using wl_egl_window_resize(). It is critical that you do *not* draw anything, nor query anything about the size/buffer age/... between `eglSwapBuffer()` and `wl_egl_window_resize()`, because EGL implementations allocate the next buffer on the first draw call or when e.g. the buffer age or size is queried. If you accidentally cause it to allocate a buffer before you resize it, the new size will be *after* the next eglSwapBuffers(), not the one you will be drawing. IIRC Vulkan is more hands on, so you don't have this issue, but on EGL, this is something to keep in mind.

**Comment by kovidgoyal** [comment]
On Thu, Oct 27, 2022 at 06:08:32AM -0700, Jonas Ådahl wrote:
> There are a couple of possibly badly documented 'gotchas" here too: with EGL, you resize the EGLSurface using wl_egl_window_resize(). It is critical that you do *not* draw anything, nor query anything about the size/buffer age/... between `eglSwapBuffer()` and `wl_egl_window_resize()`, because EGL implementations allocate the next buffer on the first draw call or when e.g. the buffer age or size is queried. If you accidentally cause it to allocate a buffer before you resize it, the new size will be *after* the next eglSwapBuffers(), not the one you will be drawing. IIRC Vulkan is more hands on, so you don't have this issue, but on EGL, this is something to keep in mind.

Yeah, another one I thought of is that kitty currently does not
render/swap until it receives a render frame from the compositor, but
to request a render frame one has to commit the surface, which cannot be
done since we are waiting for a swap to attach the correct buffer to the
surface. Thankfully in Wayland kitty already sets the swap interval to
zero to prevent EGL from requesting render frames.

I suppose I will have to have glfw return a "fake" render frame to kitty in the
case when it there is a pending surface commit.

This is going to be uber painful.

**Comment by real-or-random** [comment]
Unfortunately, this is still an issue under certain circumstances. :/ I can reproduce it only when using multiple windows in a single instance (`-1`). While your fix did something, I think the bug never completely disappeared for me. But only today, I figured out that multiple windows play a role here.

Here's how I can reproduce this reliable on mutter 45.3:
 - Set `hide_window_decorations yes`
 - Open two windows using `kitty -1 --start-as normal` (e.g., use `Alt+F2` to run commands)
 - Maximize one of the windows (press `Super+Up`), leave the other one unmaximized
 - Try to switch between the windows using `Alt+Esc`:
  - Hold `Alt` and press `Esc` a few times: The border which indicates the window to focus does strange things
  - If you release `Alt` when on the unmaximized window, the maximized window becomes unmaximized again
 
I can provide a screencast if you can't reproduce it.

I hope https://github.com/kovidgoyal/kitty/commit/2435a8ccfd17e800db7f4c54f044125510eb51b8 has fixed the root cause (thank you for taking the time!) and this is easier to fix this time. Maybe kitty just somehow mixes up the two windows.

**Comment by kovidgoyal** [comment]
Sorry, my patience with GNOME+Wayland is at an end, but patches are welcome.

**Comment by kovidgoyal** [comment]
@real-or-random this should now be fixed, at least I cannot replicate it. Try a nightly build or master. There was a bunch of wayland work, including support for fractional scaling and some changes related to that appear to have fixed it.

**Comment by real-or-random** [comment]
Yep, I can't reproduce this on master (and yes, I happen to use fractional scaling, not sure if this was related or not). Awesome, thanks!
